hkmc2/shared/src/test/mlscript-compile/Rendering.mls

```diff
@@ -35,27 +35,350 @@ fun interleave(sep)(...args) =
+  fun renderArray(array, level, keyLength, startPos) =
+    let
+      i = 0
+      length = array.length
+      emptyItemCount = 0
+      // The length of indentations that should be prepended to items in this
+      // array if this array is displayed in multi-lines.
+      itemIndentationLength =
+        if indent is null then 0 else (level + 1) * indent.length
+    // Return `true` if the array has no remaining elements.
+    fun done() =
+      let isDone = null
+      while isDone is null do
+        if...
+        i >= length then set isDone = true
+        Reflect.has(array, i) then set isDone = false
+        else
+          set { emptyItemCount += 1, i += 1 }
+      (emptyItemCount === 0) && (if isDone is null then true else isDone)
+    fun next(isFirst: Bool, prefixLength: Int): Str = if
+      emptyItemCount > 0 then
+        let emptyItemCount' = emptyItemCount
+        set emptyItemCount = 0
+        "<" + emptyItemCount' + " empty items>"
+      i < length then
+        let
+          // The starting column if the current item is placed on the same line.
+          prefixLengthIfSameLine = prefixLength + (if isFirst then 0 else 2)
+          valueStartLength = if prefixLengthIfSameLine > breakLength
+            then itemIndentationLength else prefixLengthIfSameLine
+          valueStr = renderValue(array.[i], level + 1, 0, valueStartLength)
+        set i += 1
+        valueStr
+      else throw new Error("All items in this array has been rendered.")
+    renderSequence of done, next, level, keyLength, startPos, symbolsForArray
+  fun renderSet(theSet, level, keyLength, startPos) =
+    let
+      iterator = theSet.(Symbol.iterator)()
+      peek = iterator.next()
+      // The length of indentations that should be prepended to items in this
+      // array if this array is displayed in multi-lines.
+      itemIndentationLength =
+        if indent is null then 0 else (level + 1) * indent.length
+    // Return `true` if the array has no remaining elements.
+    fun done() = peek.done
+    fun next(isFirst: Bool, prefixLength: Int): Str =
+      let result = if peek is (done: false, :value) then 
+        let
+          // The starting column if the current item is placed on the same line.
+          prefixLengthIfSameLine = prefixLength + (if isFirst then 0 else 2)
+          valueStartLength = if prefixLengthIfSameLine > breakLength
+            then itemIndentationLength else prefixLengthIfSameLine
+        renderValue(value, level + 1, 0, valueStartLength)
+      set peek = iterator.next()
+      result
+    renderSequence of
+      done, next, level, keyLength, startPos, symbolsForSet(theSet.size)
+  fun renderSequence(done, next, level: Int, keyLength: Int, startPos: Int, symbols) =
+    let
+      items = []
+      longestItem = 0
+      hasMultiline = false
+      // Try calculating the length needed to place all items in one line. The
+      // brackets and spaces around items are excluded from the length.
```

Comment:
Why are they excluded?

Reply:
This comment is outdated. I will review all the comments.